File: src/typescript/src/validation/ValidationContext.ts

import { ValidationError } from "./ValidationError";
export type JsonObject = Record<string, any>;
export class ValidationContext {
  public readonly types: Record<string, JsonObject> = {};
  public readonly errors: ValidationError[] = [];
  constructor(public readonly document: JsonObject) {
    if (document.types && typeof document.types === "object") {
      for (const [name, type] of Object.entries(document.types)) {
        if (type && typeof type === "object") this.types[name] = type;
      }
    }
  }
  error(path: string, code: string, message: string): void {
    this.errors.push(new ValidationError(path, code, message));
  }
}

File: packages/validator-cs/OpenSchema/ValidationContext.cs

using System.Text.Json.Nodes;

namespace OpenSchema.Validation;

internal sealed class ValidationContext
{
    public JsonObject Document { get; }
    public Dictionary<string, JsonObject> Types { get; } = new();
    public List<ValidationError> Errors { get; } = new();

    public ValidationContext(JsonObject document)
    {
        Document = document;
        if (document["types"] is JsonObject typesObj)
        {
            foreach (var kv in typesObj)
            {
                if (kv.Value is JsonObject tobj)
                {
                    Types[kv.Key] = tobj;
                }
            }
        }
    }

    public void Error(string path, string code, string message) => Errors.Add(new ValidationError(path, code, message));

    public static string Jp(params string[] parts) => string.Join('/', parts.Select(p => p.Replace("~", "~0").Replace("/", "~1")));
}
